### PR TITLE
MM-16872 - Extend Plugin API to set LHS bot icon

### PR DESCRIPTION
**Pull request by ali-farooq0**
#### Summary
- Extended PluginAPI to include `GetBotIconImage` `SetBotIconImage` `DeleteBotIconImage`

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-16872

Created a HW ticket to override LHS bot icon using the new plugin api
https://mattermost.atlassian.net/browse/MM-16903

### CODE REVIEW

File: app/plugin_api.go [hunk 1]
```diff
@@ -756,3 +756,38 @@ func (api *PluginAPI) UpdateBotActive(userId string, active bool) (*model.Bot, *
...
 	return api.app.PermanentDeleteBot(userId)
 }
+
+func (api *PluginAPI) GetBotIconImage(userId string) ([]byte, *model.AppError) {
+	bot, err := api.app.GetBot(userId, true)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := api.app.GetBotIconImage(bot.UserId)
```
Comment (lieut-data) [inline app/plugin_api.go]: nit: just return?

File: app/plugin_api.go [hunk 2]
```diff
@@ -756,3 +756,38 @@ func (api *PluginAPI) UpdateBotActive(userId string, active bool) (*model.Bot, *
...
+	return data, err
+}
+
+func (api *PluginAPI) SetBotIconImage(userId string, data []byte) *model.AppError {
+	bot, err := api.app.GetBot(userId, true)
+	if err != nil {
+		return err
+	}
+
+	fileReader := bytes.NewReader(data)
```
Comment (lieut-data) [inline app/plugin_api.go]: nit: maybe just inline?

File: app/plugin_api.go [hunk 3]
```diff
@@ -756,3 +756,38 @@ func (api *PluginAPI) UpdateBotActive(userId string, active bool) (*model.Bot, *
 func (api *PluginAPI) PermanentDeleteBot(userId string) *model.AppError {
 	return api.app.PermanentDeleteBot(userId)
 }
+
+func (api *PluginAPI) GetBotIconImage(userId string) ([]byte, *model.AppError) {
+	bot, err := api.app.GetBot(userId, true)
```
Comment (lieut-data) [inline app/plugin_api.go]: A bot's user id is equivalent to the id of the user backing the bot -- so it should suffice to just pass `userId` through to `app.GetBotIconImage`, `app.SetBotIconImage`, etc.

However, I note that `app.GetBotIconImage` doesn't currently return an error if it's not a bot -- should it?

Comment (ali-farooq0) [reply app/plugin_api.go]: In the api handler i was fetching the user and checking there as I need to use a user field, i thought fetching user twice might put extra load on the `get` api as it'll happen more frequent. But i think it should be ok. Better to be consistent across the 3 calls to verify if a user is a bot.

File: app/bot.go [hunk 4]
```diff
@@ -247,17 +262,21 @@ func (a *App) DeleteBotIconImage(botUserId string) *model.AppError {
 }
 
 // GetBotIconImage retrieves LHS icon for a bot.
-func (a *App) GetBotIconImage(botUserId string) ([]byte, bool, *model.AppError) {
+func (a *App) GetBotIconImage(botUserId string) ([]byte, *model.AppError) {
+	if _, err := a.GetBot(botUserId, true); err != nil {
+		return nil, err
+	}
+
 	if len(*a.Config().FileSettings.DriverName) == 0 {
```
Comment (marianunez) [inline app/bot.go]: ☝️Same comment as above regarding the filestore check

File: app/bot.go [hunk 5]
```diff
@@ -197,19 +198,29 @@ func (a *App) ConvertUserToBot(user *model.User) (*model.Bot, *model.AppError) {
...
+	return a.SetBotIconImage(botUserId, file)
+}
+
+// SetBotIconImage sets LHS icon for a bot.
+func (a *App) SetBotIconImage(botUserId string, file io.ReadSeeker) *model.AppError {
+	if _, err := a.GetBot(botUserId, true); err != nil {
+		return err
+	}
+
+	if len(*a.Config().FileSettings.DriverName) == 0 {
```
Comment (marianunez) [inline app/bot.go]: This check is not necessary at this level given that when you use instantiate a filestore it does check for `DriverName` and will return an appropriate error [here](https://github.com/mattermost/mattermost-server/blob/5690df9d959078bb1017dfd7e36abc4274bd890f/services/filesstore/filesstore.go#L52) if not configured.

**Pull request merged**